ceno_zkvm/src/instructions/riscv/mul.rs

```diff
@@ -246,21 +278,27 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for MulhInstructionBas
 
         // 2. Verify that the product of signed inputs `rs1` and `rs2` is equal to
         //    the result of interpreting `rd` as the high limb of a 2s complement
-        //    value with some 32-bit low limb
-
-        let prod_low = UInt::new(|| "prod_low", circuit_builder)?;
-        circuit_builder.require_equal(
-            || "validate_prod_high_limb",
-            rs1_val * rs2_val,
-            rd_val * (1u64 << 32) + prod_low.value(),
-        )?;
+        //    value
+        if I::INST_KIND == InsnKind::MUL {
+            circuit_builder.require_equal(
+                || "validate_prod_low_limb",
+                rs1_val * rs2_val,
+                prod_half.value() * (1u64 << BIT_WIDTH) + rd_val,
+            )?;
+        } else {
+            circuit_builder.require_equal(
+                || "validate_prod_high_limb",
+                rs1_val * rs2_val,
+                rd_val * (1u64 << BIT_WIDTH) + prod_half.value(),
```

Comment:
Similar here:

```suggestion
                (rd_val  << BIT_WIDTH) + prod_half.value(),
```

Reply:
the same case, it's an `Expression`.

Reply:
That works now, see https://github.com/scroll-tech/ceno/pull/663

Reply:
fixed in [2114670](https://github.com/scroll-tech/ceno/pull/660/commits/2114670c5281259da340908eddecb07f29d4b608)